File: services/alchemy-api/src/current_alchemy/api/policy.py
"""Authoritative endpoint cache and edge-rate policy registry."""

from dataclasses import dataclass
from enum import StrEnum
from re import Pattern, compile


class EndpointClass(StrEnum):
    PUBLIC_CACHEABLE = "public-cacheable"
    PUBLIC_UNCACHEABLE = "public-uncacheable"
    PRIVATE_NO_STORE = "private-no-store"
    HEALTH = "health"
    ADMINISTRATIVE = "administrative"


class RatePolicyClass(StrEnum):
    ANONYMOUS_PUBLIC_READ = "anonymous-public-read"
    AUTHENTICATED_READ = "authenticated-read"
    SEARCH = "search"
    FORMULA_ANALYSIS = "formula-analysis"
    FUTURE_INTELLIGENCE = "future-intelligence"
    ADMINISTRATIVE_IMPORT = "administrative-import"


@dataclass(frozen=True, slots=True)
class EndpointPolicy:
    endpoint_class: EndpointClass
    rate_class: RatePolicyClass
    cache_control: str = "no-store"


@dataclass(frozen=True, slots=True)
class RegisteredPolicy:
    methods: frozenset[str]
    path: Pattern[str]
    policy: EndpointPolicy


NO_STORE = EndpointPolicy(
    EndpointClass.PUBLIC_UNCACHEABLE,
    RatePolicyClass.ANONYMOUS_PUBLIC_READ,
)
PRIVATE_NO_STORE = EndpointPolicy(
    EndpointClass.PRIVATE_NO_STORE,
    RatePolicyClass.AUTHENTICATED_READ,
    "private, no-store",
)
HEALTH = EndpointPolicy(EndpointClass.HEALTH, RatePolicyClass.ANONYMOUS_PUBLIC_READ)
ADMINISTRATIVE = EndpointPolicy(
    EndpointClass.ADMINISTRATIVE,
    RatePolicyClass.ADMINISTRATIVE_IMPORT,
    "private, no-store",
)
SEARCH = EndpointPolicy(
    EndpointClass.PUBLIC_CACHEABLE,
    RatePolicyClass.SEARCH,
    "public, max-age=0, s-maxage=60, stale-while-revalidate=300",
)
STABLE_RECORD = EndpointPolicy(
    EndpointClass.PUBLIC_CACHEABLE,
    RatePolicyClass.ANONYMOUS_PUBLIC_READ,
    "public, max-age=60, s-maxage=3600, stale-while-revalidate=86400",
)
FORMULA_ANALYSIS = EndpointPolicy(
    EndpointClass.PUBLIC_UNCACHEABLE,
    RatePolicyClass.FORMULA_ANALYSIS,
)
FUTURE_INTELLIGENCE = EndpointPolicy(
    EndpointClass.PRIVATE_NO_STORE,
    RatePolicyClass.FUTURE_INTELLIGENCE,
    "private, no-store",
)


_REGISTRY: tuple[RegisteredPolicy, ...] = (
    RegisteredPolicy(frozenset({"GET"}), compile(r"^/api/v1/health/(live|ready)$"), HEALTH),
    RegisteredPolicy(
        frozenset({"GET", "POST", "PUT", "PATCH", "DELETE"}),
        compile(r"^/api/v1/(admin|internal|imports)(?:/.*)?$"),
        ADMINISTRATIVE,
    ),
    RegisteredPolicy(
        frozenset({"GET", "POST", "PUT", "PATCH", "DELETE"}),
        compile(r"^/api/v1/(auth|private|users|profiles|subscriptions|memories)(?:/.*)?$"),
        PRIVATE_NO_STORE,
    ),
    RegisteredPolicy(
        frozenset({"GET", "POST"}),
        compile(r"^/api/v1/(intelligence|inquiry)(?:/.*)?$"),
        FUTURE_INTELLIGENCE,
    ),
    RegisteredPolicy(
        frozenset({"POST"}),
        compile(r"^/api/v1/formulas/(analyze|compare)$"),
        FORMULA_ANALYSIS,
    ),
    RegisteredPolicy(
        frozenset({"POST"}),
        compile(r"^/api/v1/(explore/query|retrieval/context)$"),
        PRIVATE_NO_STORE,
    ),
    RegisteredPolicy(
        frozenset({"GET"}),
        compile(r"^/api/v1/(meta|sources|documents|passages|graph/entities/[^/]+/neighborhood)$"),
        STABLE_RECORD,
    ),
    RegisteredPolicy(
        frozenset({"GET"}),
        compile(r"^/api/v1/(search/suggest|text/search|herbs|formulas)$"),
        SEARCH,
    ),
    RegisteredPolicy(
        frozenset({"GET"}),
        compile(r"^/api/v1/(herbs|formulas|sources|documents|passages)/[^/]+$"),
        STABLE_RECORD,
    ),
)


def endpoint_policy(method: str, path: str) -> EndpointPolicy:
    """Return a deny-by-default policy for one HTTP method and path."""

    normalized_method = method.upper()
    for registration in _REGISTRY:
        if normalized_method in registration.methods and registration.path.fullmatch(path):
            return registration.policy
    return NO_STORE


def registered_policies() -> tuple[RegisteredPolicy, ...]:
    """Expose the immutable registry to documentation and contract tests."""

    return _REGISTRY

File: workers/api-gateway/src/policy.ts
export const API_PREFIX = '/api/v1/'
export const PROMPT_LAB_PREFIX = '/api/gene-keys-lab/'
export const MAX_REQUEST_BODY_BYTES = 1_048_576

export type EndpointClass =
  'public-cacheable' | 'public-uncacheable' | 'private-no-store' | 'health' | 'administrative'

export type RatePolicyClass =
  | 'anonymous-public-read'
  | 'authenticated-read'
  | 'search'
  | 'formula-analysis'
  | 'graph-retrieval'
  | 'future-intelligence'
  | 'administrative-import'

export interface RoutePolicy {
  endpointClass: EndpointClass
  rateClass: RatePolicyClass
}

const HEALTH_PATHS = new Set(['/api/v1/health/live', '/api/v1/health/ready'])
const PRIVATE_PREFIXES = [
  '/api/v1/auth/',
  '/api/v1/private/',
  '/api/v1/users/',
  '/api/v1/profiles/',
  '/api/v1/subscriptions/',
  '/api/v1/memories/',
]
const ADMINISTRATIVE_PREFIXES = ['/api/v1/admin/', '/api/v1/internal/', '/api/v1/imports/']
const INTELLIGENCE_PREFIXES = ['/api/v1/intelligence/', '/api/v1/inquiry/']

const stableRecord = /^\/api\/v1\/(meta|sources|documents|passages)$/
const stableDetail = /^\/api\/v1\/(herbs|formulas|sources|documents|passages)\/[^/]+$/
const graphDetail = /^\/api\/v1\/graph\/entities\/[^/]+\/neighborhood$/
const search = /^\/api\/v1\/(search\/suggest|text\/search|herbs|formulas)$/
const privateBoundedPost = new Set(['/api/v1/explore/query', '/api/v1/retrieval/context'])

const hasPrefix = (path: string, prefixes: readonly string[]): boolean =>
  prefixes.some((prefix) => path === prefix.slice(0, -1) || path.startsWith(prefix))

export function routePolicy(method: string, path: string): RoutePolicy {
  if (path.startsWith(PROMPT_LAB_PREFIX)) {
    return { endpointClass: 'private-no-store', rateClass: 'future-intelligence' }
  }
  if (HEALTH_PATHS.has(path)) {
    return { endpointClass: 'health', rateClass: 'anonymous-public-read' }
  }
  if (hasPrefix(path, ADMINISTRATIVE_PREFIXES)) {
    return { endpointClass: 'administrative', rateClass: 'administrative-import' }
  }
  if (hasPrefix(path, PRIVATE_PREFIXES)) {
    return { endpointClass: 'private-no-store', rateClass: 'authenticated-read' }
  }
  if (hasPrefix(path, INTELLIGENCE_PREFIXES)) {
    return { endpointClass: 'private-no-store', rateClass: 'future-intelligence' }
  }
  if (method === 'POST' && privateBoundedPost.has(path)) {
    return { endpointClass: 'public-uncacheable', rateClass: 'graph-retrieval' }
  }
  if (method === 'POST' && /^\/api\/v1\/formulas\/(analyze|compare)$/.test(path)) {
    return { endpointClass: 'public-uncacheable', rateClass: 'formula-analysis' }
  }
  if (method === 'GET' && search.test(path)) {
    return { endpointClass: 'public-cacheable', rateClass: 'search' }
  }
  if (
    method === 'GET' &&
    (stableRecord.test(path) || stableDetail.test(path) || graphDetail.test(path))
  ) {
    return { endpointClass: 'public-cacheable', rateClass: 'anonymous-public-read' }
  }
  return { endpointClass: 'public-uncacheable', rateClass: 'anonymous-public-read' }
}

export function isPublicCacheCandidate(request: Request, policy: RoutePolicy): boolean {
  const requestCacheControl = request.headers.get('Cache-Control') ?? ''
  return (
    request.method === 'GET' &&
    policy.endpointClass === 'public-cacheable' &&
    !request.headers.has('Authorization') &&
    !request.headers.has('Cookie') &&
    !request.headers.has('Range') &&
    !requestCacheControl.includes('no-cache') &&
    !requestCacheControl.includes('no-store') &&
    request.headers.get('Pragma') !== 'no-cache'
  )
}

export function declaredBodyTooLarge(request: Request): boolean {
  const value = request.headers.get('Content-Length')
  if (value === null) return false
  const bytes = Number(value)
  return !Number.isSafeInteger(bytes) || bytes < 0 || bytes > MAX_REQUEST_BODY_BYTES
}
